graphql/schemabuilder/batch.go

```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		return nil, nil, in, fmt.Errorf("attempted to parse %s as arguments struct, but failed: %s", inType.Name(), err.Error())
+	}
+	inputObject, ok := argType.(*graphql.InputObject)
+	if !ok {
+		return nil, nil, nil, fmt.Errorf("%s's args should be an object", funcCtx.funcType)
+	}
+	args := make(map[string]graphql.Type, len(inputObject.InputFields))
+	for name, typ := range inputObject.InputFields {
+		args[name] = typ
+	}
+	funcCtx.hasArgs = true
+	return argParser, args, in, nil
+}
+
+// consumeSelectionSet reads the input parameters and will pop off the
+// selectionSet type if we detect it in the input fields.  Check out
+// graphql.SelectionSet for more infomation about selection sets.
+func (funcCtx *batchFuncContext) consumeSelectionSet(in []reflect.Type) []reflect.Type {
+	if len(in) > 0 && in[0] == selectionSetType {
+		in = in[1:]
+		funcCtx.hasSelectionSet = true
+	}
+	return in
+}
+
+func (funcCtx *batchFuncContext) getFuncOutputTypes() []reflect.Type {
+	out := make([]reflect.Type, 0, funcCtx.funcType.NumOut())
+	for i := 0; i < funcCtx.funcType.NumOut(); i++ {
+		out = append(out, funcCtx.funcType.Out(i))
+	}
+	return out
+}
+
+// consumeReturnValue consumes the function output's response value if it exists
+// and validates that the response is a proper batch type.
+func (funcCtx *batchFuncContext) consumeReturnValue(m *method, sb *schemaBuilder, out []reflect.Type) (graphql.Type, []reflect.Type, error) {
+	if len(out) == 0 || out[0] == errType {
+		if m.MarkedNonNullable {
+			return nil, nil, fmt.Errorf("%s is marked non-nullable, but has no return value", funcCtx.funcType)
+		}
+		retType, err := sb.getType(reflect.TypeOf(true))
+		if err != nil {
+			return nil, nil, err
+		}
+		return retType, out, nil
+	}
+	outType := out[0]
+	out = out[1:]
+	if outType.Kind() != reflect.Map ||
+		outType.Key().Kind() != reflect.Int {
+		return nil, nil, fmt.Errorf(
+			"invalid response batch type, expected map[int]<Type>, but got %s",
+			outType.String(),
+		)
+	}
+	retType, err := sb.getType(outType.Elem())
+	if err != nil {
+		return nil, nil, err
+	}
+	if m.MarkedNonNullable { // TODO see if we can even enforce non-nil resps for batch endpoints
+		if _, ok := retType.(*graphql.NonNull); !ok {
+			retType = &graphql.NonNull{Type: retType}
+		}
+	}
+	funcCtx.hasRet = true
+	return retType, out, nil
+}
+
+// consumeReturnValue consumes the function output's error type if it exists.
+func (funcCtx *batchFuncContext) consumeReturnError(out []reflect.Type) []reflect.Type {
+	if len(out) > 0 && out[0] == errType {
+		funcCtx.hasError = true
+		out = out[1:]
+	}
+	return out
+}
+
+// prepareResolveArgs converts the provided source, args and context into the
+// required list of reflect.Value types that the function needs to be called.
+func (funcCtx *batchFuncContext) prepareResolveArgs(sources []interface{}, args interface{}, ctx context.Context, selectionSet *graphql.SelectionSet) []reflect.Value {
```

Comment:
i'd make this a function instead of a method because i read context as descriptor now and i'd use a descriptor but not ask it to do things... if that makes sense